pulpcore/app/models/content.py

```
@@ -560,21 +561,24 @@ class SigningService(BaseModel):
 
     Fields:
         name (models.TextField):
-            A unique name describing the script (or executable) used for signing.
+            A unique name describing a script (or executable) used for signing.
+        public_key (models.TextField):
+            An absolute path to a public key.
         script (models.TextField):
-            An absolute path to the external signing script (or executable).
+            An absolute path to an external signing script (or executable).
 
     """
 
     name = models.TextField(db_index=True, unique=True)
+    public_key = models.TextField(default="")
```

I imagined this would be a required field. All SigningServices use PKI which use public keys is what I was thinking.

---

I think this is a migration thing. What about existing signing_services?
IIRC django won't even let you add a required field without providing a default.

---

With 3.9 signing servies are immutable so I think we've already gotten to the point where users will need to make them new.

---

So yes to required. Can we also not default it to anything and provide a more verbose default only for the migration?

---

We could provide a migration which sets the value at migration time. It could sidestep the hook by calling https://rsinger86.github.io/django-lifecycle/advanced/#suppressing-hooked-methods in the data migration itself. What do you think about that idea?

---

I thought django would ask you for a migration default if you don't have it on the model.

---

I was imaging a migraiton that has 3 parts:

1. Add the field but make it optional
2. Perform the data migration populating the field
3. SQL update to make that field required (now that all existing data also has it). This last part would "match" the model code marking the field as required.

---

I do not quite understand what are you trying to achieve with those 3 parts, @bmbouter. How do you want to populate the data from an old SigningService object if there was not any `public_key` present back then? It will contain the same data like before. Do you want to extract the value of the public key by executing a signing script and manually reading the value of `return_value["key"]` within the migration?

---

> I do not quite understand what are you trying to achieve with those 3 parts, @bmbouter. How do you want to populate the data from an old SigningService object if there was not any `public_key` present back then? It will contain the same data like before. Do you want to extract the value of the public key by executing a signing script and manually reading the value of `return_value["key"]` within the migration?

That's what I was thinking yes. If the migration fails, I think it should probably put in dummy data and log to the user an error and print the error to stderr as well so it's shown in the foreground too. We can't have the migraitons "fail to apply"

@mdellweg what do you think about ^?

---

Agreed, we must make the migration apply cleanly at any cost.
Trying to extract the public key is a convenience we offer the user, but it could fail for various reasons (revoked/expired keys is probably just one of them).
So i think, we should make the failing units unusable somehow to force the user to reinstanciate them. Having dummy data in the key field breaks the promise we give our plugins IMHO.

---

@mdellweg agreed. In terms of making the failing units unusable, I was thinking it would play out like:

1. The migration logs them and puts an empty string in the public_key field.
2. Then the SQL alter to make public_key required can be applied without error and the transaction can finish.
3. The code expects the public_key data to be accurate and so in cases where it's just `''` the code will receive a RuntimeError. Or perhaps it could even check if its '' and gracefully refuse. Either way is ok with me.